core/src/main/scala/org/bykn/bosatsu/TypedExprNormalization.scala

```diff
@@ -374,22 +389,39 @@ object TypedExprNormalization {
         }
       }
 
-    case class WithScope[A](scope: Scope[A]) {
+    case class WithScope[A](scope: Scope[A], typeEnv: TypeEnv[Kind.Arg]) {
+      private lazy val kindOf: Type => Option[Kind] =
+        { case const @ Type.TyConst(_) =>
+            typeEnv.getType(const).map(_.kindOf)
+          case _ => None
+        }
+
       object ResolveToLambda {
-        def unapply(te: TypedExpr[A]): Option[(NonEmptyList[(Bindable, Type)], TypedExpr[A], A)] =
+        def unapply(te: TypedExpr[A]): Option[(List[(Type.Var.Bound, Kind)], NonEmptyList[(Bindable, Type)], TypedExpr[A], A)] =
           te match {
-            case AnnotatedLambda(args, expr, ltag) => Some((args, expr, ltag))
+            case Annotation(ResolveToLambda((h :: t), args, ex, tag), rho: Type.Rho) =>
```

Comment:
these changes make function inlining much more powerful, which turns out to blow a lot of things up.

With unconditional function inlining, we can definitely make code exponentially large.

It either caused some issues with infinite loops, or maybe just an exponential blow up that was blowing the stack.

I tried to hack around it by limiting symbolic evaluation to when it can make expressions smaller, but that broke a bunch of other tests.